samza-api/src/main/java/org/apache/samza/operators/MessageStream.java

```diff
@@ -16,173 +16,109 @@
    */
-  public MessageStream<M> filter(Function<M, Boolean> filter) {
-    Operator<M> op = Operators.<M, M>getStreamOperator(t -> new ArrayList<M>() { {
-        if (filter.apply(t)) {
-          this.add(t);
-        }
-      } });
-    this.subscribers.add(op);
-    return op.getOutputStream();
-  }
+  MessageStream<M> filter(FilterFunction<M> filterFn);
 
   /**
-   * Method to send an input {@link MessageStream} to an output {@link org.apache.samza.system.SystemStream}, and allows the output {@link MessageStream}
-   * to be consumed by downstream stream operators again.
+   * Allows sending {@link Message}s in this {@link MessageStream} to an output
+   * {@link org.apache.samza.system.SystemStream} using the provided {@link SinkFunction}.
    *
-   * @param sink  the user-defined sink function to send the input {@link Message}s to the external output systems
+   * This is a terminal transform and does not allow chaining further transforms.
+   *
+   * @param sinkFn  the sink function to send {@link Message}s in this stream to output systems
    */
-  public void sink(VoidFunction3<M, MessageCollector, TaskCoordinator> sink) {
-    this.subscribers.add(Operators.getSinkOperator(sink));
-  }
+  void sink(SinkFunction<M> sinkFn);
 
   /**
-   * Method to perform a window function (i.e. a group-by, aggregate function) on a {@link MessageStream}
+   * Windows this {@link MessageStream} according to the provided {@link Window} semantics (e.g. tumbling,
+   * sliding or session windows) and sends window results to the output {@link MessageStream}.
+   *
+   * Use the {@link org.apache.samza.operators.windows.Windows} helper methods to create the appropriate windows.
    *
-   * @param window  the window function to group and aggregate the input {@link Message}s from the input {@link MessageStream}
-   * @param <WK>  the type of key in the output {@link Message} from the {@link Windows.Window} function
-   * @param <WV>  the type of output value from
-   * @param <WS>  the type of window state kept in the {@link Windows.Window} function
-   * @param <WM>  the type of {@link org.apache.samza.operators.internal.WindowOutput} message from the {@link Windows.Window} function
-   * @return the output {@link MessageStream} after applying the window function on the input {@link MessageStream}
+   * @param window  the {@link Window} to group and process the input {@link Message}s from the input {@link MessageStream}
+   * @param <WK>  the type of key in the output {@link Message} from the {@link Window} function
+   * @param <WV>  the type of output value from the {@link Window} function
+   * @param <WS>  the type of window state kept in the {@link Window} function
+   * @param <WM>  the type of {@link WindowOutput} {@link Message} from the {@link Window} function
+   * @return  the output {@link MessageStream} after applying the window function on the input {@link MessageStream}
```

Comment:
Wasn't sure if it's still going to be a public API after your changes. Maybe change to the generic "the type of value in the ouput {@link Message} in the output {@link MessageStream}"?

On second thought, maybe I should replace "output" with "transformed" everywhere?